src/registryDAO.mligo

```diff
@@ -54,34 +55,44 @@ let apply_diff_affected
  *   3.b. is a token transfer
  * where min_xtz_amount and max_xtz_amount are from storage configuration.
  *)
-let registry_DAO_proposal_check (params, extras : propose_params * contract_extra) : bool =
+let registry_DAO_proposal_check (params, extras : propose_params * contract_extra) : unit =
   let proposal_size = Bytes.size(params.proposal_metadata) in
   let frozen_scale_value = unpack_nat(find_big_map("frozen_scale_value", extras)) in
   let frozen_extra_value = unpack_nat(find_big_map("frozen_extra_value", extras)) in
   let max_proposal_size = unpack_nat(find_big_map("max_proposal_size", extras)) in
 
   let required_token_lock = frozen_scale_value * proposal_size + frozen_extra_value in
-  let has_correct_token_lock =
-    (params.frozen_token = required_token_lock) && (proposal_size < max_proposal_size) in
-
-  if has_correct_token_lock then
-    match unpack_proposal_metadata(params.proposal_metadata) with
-    | Transfer_proposal tp ->
-        let min_xtz_amount = unpack_tez(find_big_map("min_xtz_amount", extras)) in
-        let max_xtz_amount = unpack_tez(find_big_map("max_xtz_amount", extras)) in
-        let is_all_transfers_valid (is_valid, transfer_type: bool * transfer_type) =
-          match transfer_type with
-          | Token_transfer_type _tt -> is_valid
-          | Xtz_transfer_type xt ->
-                 is_valid && xt.amount <> 0mutez
-              && min_xtz_amount <= xt.amount && xt.amount <= max_xtz_amount
-        in
-        List.fold is_all_transfers_valid tp.transfers has_correct_token_lock
-    | Update_receivers_proposal _urp -> has_correct_token_lock
-    | Configuration_proposal _cp -> has_correct_token_lock
-
-  else
-    false
+  let pm = unpack_proposal_metadata(params.proposal_metadata) in
+
+  let _ : unit =
+    if (params.frozen_token <> required_token_lock) then
+      fail_proposal_check (incorrect_token_amount_err_msg)
+    else if (proposal_size >= max_proposal_size) then
+      fail_proposal_check (too_large_proposal_err_msg)
+    else
+      unit in
+
+  match unpack_proposal_metadata(params.proposal_metadata) with
+  | Transfer_proposal tp ->
+      let min_xtz_amount = unpack_tez(find_big_map("min_xtz_amount", extras)) in
+      let max_xtz_amount = unpack_tez(find_big_map("max_xtz_amount", extras)) in
+      let is_all_transfers_valid (transfer_type: transfer_type) =
+        match transfer_type with
+        | Token_transfer_type _tt -> unit
+        | Xtz_transfer_type xt ->
+            if (xt.amount = 0mutez) then
+              fail_proposal_check (zero_mutez_err_msg)
+            else if (xt.amount < min_xtz_amount) then
+              fail_proposal_check (too_small_xtz_err_msg)
+            else if (xt.amount > max_xtz_amount) then
+              fail_proposal_check (too_large_xtz_err_msg)
+            else
+              unit
```

Comment:
One more suggestion about these multi-branch if: this implementation is now a bit too expensive as it's based on multiple `fail_proposal_check` this means that there are 2 `PUSH` , 1 `PAIR` and a `FAILWITH` for every branch.

What I suggest is to move the checking to another function that returns a `string option`, e.g. in this case:
```ocaml
let check_xtz_transfer (xt, min_xtz_amount, max_xtz_amount : xtz_transfer * tez * tez) : string option =
  if (xt.amount = 0mutez) then Some zero_mutez_err_msg
  else if (xt.amount < min_xtz_amount) then Some too_small_xtz_err_msg
  else if (xt.amount > max_xtz_amount) then Some too_large_xtz_err_msg
  else None
```
and in turn only call `fail_proposal_check` in case this returns `Some` error message, here:
```suggestion
        | Xtz_transfer_type xt ->
            begin
              match check_xtz_transfer(xt, min_xtz_amount, max_xtz_amount) with
              | Some err_msg -> fail_proposal_check(err_msg)
              | None -> unit
            end
```

One more thing is: in this case you may notice that the check itself is common between `registry` nad `treasury`, so `check_xtz_transfer` could also be moved in the common module, I think.

Reply:
I've add `check_xtz_transfer` to a common file, and update the proposal_check now.